src/target/riscv/riscv.c

```diff
@@ -700,6 +734,110 @@ static void log_trigger_request_info(struct trigger_request_info trig_info)
+	elem->tdata1 = tdata1;
+	INIT_LIST_HEAD(&elem->tdata2_cache_head);
+	list_add_tail(&elem->elem_tdata1, tdata1_cache_head);
+	return elem;
+}
+
+static void create_new_tdata2_cache(struct list_head *tdata2_cache_head, riscv_reg_t tdata2)
+{
+
+	struct tdata2_cache * const elem = calloc(1, sizeof(struct tdata2_cache));
+	elem->tdata2 = tdata2;
+	list_add(&elem->elem_tdata2, tdata2_cache_head);
+}
+
+struct tdata2_cache *find_call_in_tdata2_cache(struct list_head *tdata2_cache_head, riscv_reg_t find_tdata2)
+{
+
+	struct tdata2_cache *elem_2;
+	list_for_each_entry(elem_2, tdata2_cache_head, elem_tdata2) {
+		if (elem_2->tdata2 == find_tdata2) {
+			return elem_2;
+		}
+	}
+	return NULL;
+}
+
+struct tdata1_cache *find_tdata1(struct list_head *tdata1_cache_head, riscv_reg_t find_tdata1)
+{
+
+	struct tdata1_cache *elem_1;
+	list_for_each_entry(elem_1, tdata1_cache_head, elem_tdata1) {
+		if (elem_1->tdata1 == find_tdata1)
+			return elem_1;
+	}
+	return NULL;
+}
+
+static void create_wp_trigger_cache(struct target *target)
+{
+
+	RISCV_INFO(r);
+
+	r->wp_triggers_cache = (struct list_head *)calloc(r->trigger_count,
+		sizeof(struct list_head));
+	for (unsigned int i = 0; i < r->trigger_count; ++i) {
+		INIT_LIST_HEAD(&r->wp_triggers_cache[i]);
+	}
+}
+
+static void add_use_wp_trigger_in_cache(struct target *target, unsigned int idx, riscv_reg_t tdata1,
+	riscv_reg_t tdata2, int error_code)
+{
+
+	RISCV_INFO(r);
+	
+	if (error_code != ERROR_TARGET_RESOURCE_NOT_AVAILABLE)
+		return;
+
+	struct tdata1_cache *tdata1_cache = find_tdata1(&(r->wp_triggers_cache[idx]), tdata1);
+	if (tdata1_cache == NULL) {
+		tdata1_cache = create_new_tdata1_cache(&(r->wp_triggers_cache[idx]), tdata1);
+	} else {
+		struct tdata2_cache *tdata2_cache = find_call_in_tdata2_cache(&(tdata1_cache->tdata2_cache_head), tdata2);
+		if (tdata2_cache != NULL) {
+			list_move(&(tdata2_cache->elem_tdata2), &(tdata1_cache->tdata2_cache_head));
+			return;
+		}
+	}
+	create_new_tdata2_cache(&(tdata1_cache->tdata2_cache_head), tdata2);
+}
+
+static bool is_use_wp_trigger_in_cache(struct target *target, unsigned int idx,
+	riscv_reg_t tdata1, riscv_reg_t tdata2)
+{
+
+	RISCV_INFO(r);
+
+	struct tdata1_cache *tdata1_cache = find_tdata1(&(r->wp_triggers_cache[idx]), tdata1);
+	if (tdata1_cache == NULL) {
+		return false;
+	}
+	struct tdata2_cache *tdata2_cache = find_call_in_tdata2_cache(&(tdata1_cache->tdata2_cache_head), tdata2);
+	if (tdata2_cache == NULL) {
+		return false;
+	}
+	assert(tdata1_cache->tdata1 == tdata1 && tdata2_cache->tdata2 == tdata2);
+	return true;
+}
+
+static int try_use_trigger_and_cache_result(struct target *target, unsigned int idx, riscv_reg_t tdata1,
+	riscv_reg_t tdata2, riscv_reg_t tdata1_ignore_mask)
+{
+	if (is_use_wp_trigger_in_cache(target, idx, tdata1, tdata2))
+			return ERROR_TARGET_RESOURCE_NOT_AVAILABLE;
+
+	int ret = set_trigger(target, idx, tdata1, tdata2, tdata1_ignore_mask);
+	add_use_wp_trigger_in_cache(target, idx, tdata1, tdata2, ret);
```

Comment:
Myself, I would probably not handle the return value inside the add_use_wp_trigger_in_cache(). Also, I think if it returned an error which isn't ERROR_TARGET_RESOURCE_NOT_AVAILABLE (error when setting/reading triggers), the cache should be invalidated. As we have no idea at which step the error happened.

Reply:
In this case, this check is needed in order not to cache calls that have a return code of ERROR_OK

Reply:
I agree with @MarekVCodasip that the error check should be here, not inside add_use_wp_trigger_in_cache().

It also needs a comment like "Add these values to the cache to remember that they are not supported."

Reply:
Addressed.